# <div align="center"> TUGAS I LECTURE MACHINE LEARNING </div>
Nama  : Ragil Muhamad Ramadhan

NIM   : 00000058511

In [1]:
# Run this code when you restart the machine
# Fill in with YOUR name and NIM
import datetime
import uuid

myName = "Ragil Muhamad Ramadhan"
myNIM = ""

In [2]:
myDate = datetime.datetime.now()
myDevice = str(uuid.uuid1())

print("Name: \t\t{}".format(myName))
print("NIM: \t\t{}".format(myNIM))
print("Start: \t\t{}".format(myDate))
print("Device ID: \t{}".format(myDevice))

Name: 		Ragil Muhamad Ramadhan
NIM: 		
Start: 		2023-11-16 13:52:38.205100
Device ID: 	bad64abd-844c-11ee-8532-e2bea1734fd9


In [3]:
# No 1. Normalisasi Data dalam Machine Learning
X = [350, 425, 280, 310, 550, 700, 600, 800, 450, 320]

# Nilai minimun dan maksimum
min_X = 280
max_X = 800
print(f"min : {min_X}, max : {max_X}")

# Normalisasi setiap data
X_sum = 350 + 425 + 280 + 310 + 550 + 700 + 600 + 800 + 450 + 320
X_length = 10
X_mean = X_sum / X_length
normalized_data = [(X[0] - min_X) / (max_X - min_X), (X[1] - min_X) / (max_X - min_X), 
                   (X[2] - min_X) / (max_X - min_X), (X[3] - min_X) / (max_X - min_X), 
                   (X[4] - min_X) / (max_X - min_X), (X[5] - min_X) / (max_X - min_X), 
                   (X[6] - min_X) / (max_X - min_X), (X[7] - min_X) / (max_X - min_X),
                   (X[8] - min_X) / (max_X - min_X), (X[9] - min_X) / (max_X - min_X)]
print(f"\na) Data yang sudah dinormalisasi : {normalized_data}")

# Mean dan Std Dev
X_mean = X_sum / X_length
std_dev = ((((X[0] - X_mean) ** 2 )+ ((X[1] - X_mean) ** 2)  + ((X[2] - X_mean) ** 2)  
           + ((X[3] - X_mean) ** 2) + ((X[4] - X_mean) ** 2)  + ((X[5] - X_mean) ** 2)
           + ((X[6] - X_mean) ** 2) + ((X[7] - X_mean) ** 2) + ((X[8] - X_mean) ** 2)
           + ((X[9] - X_mean) ** 2)) / X_length) ** 0.5
print(f"\nmean : {X_mean}, std_dev = {std_dev}")

# Standard scalling
# Standard scalling
z_score = [(X[0] - X_mean) / std_dev , (X[1] - X_mean)  / std_dev , (X[2] - X_mean) / std_dev 
           , (X[3] - X_mean)  / std_dev , (X[4] - X_mean) / std_dev , (X[5] - X_mean) / std_dev
           , (X[6] - X_mean)  / std_dev , (X[7] - X_mean) / std_dev , (X[8] - X_mean) / std_dev
           , (X[9] - X_mean)  / std_dev]
print(f"\nb) Z-score : {z_score}")

min : 280, max : 800

a) Data yang sudah dinormalisasi : [0.1346153846153846, 0.27884615384615385, 0.0, 0.057692307692307696, 0.5192307692307693, 0.8076923076923077, 0.6153846153846154, 1.0, 0.3269230769230769, 0.07692307692307693]

mean : 478.5, std_dev = 168.93859831311494

b) Z-score : [-0.7606313849120196, -0.3166831057804907, -1.1749831121014467, -0.9974038004488351, 0.4232306927720576, 1.3111272510351155, 0.719196212193077, 1.9030582898771542, -0.16870034606998102, -0.9382106965646312]


#3. Apa perbedaan antara Min-Max Scaling dan Standard Scaling dalam konteks ini?

    Min-Max Scalling merubah data menjadi interval antara 0 - 1, sedangkan pada standard scalling melihat distribusi data apakah normal atau tidak. Pada konteks di atas min-max scallling lebih baik karena tidak ada outliers dalam data, namun pada data yang tidak terdistribusi normal penggunaan standard scalling karena dapat mengukur dan merubah distribusi menjadi normal.
    
#4. Bagaimana pemilihan metode scaling dapat memengaruhi hasil deteksi anomali dalam lalu lintas jaringan?

    Pemilihan metode dapat mempengaruhi dalam mencari anomali, dimana persebaran data terjadi dengan tidak wajar. Maka dari itu peran standard scalling menjadi sangat membantu, karena dapat melihat persebaran data terjadi secara normal maupun tidak. Sedangkan min-max scalling tidak dapat mengidentifikasi persebaran normal secara langsung, akan tetapi masih dapat dilihat selisih dari hasil normalisasi yang terjadi.

In [4]:
# 2. Jarak vektor dan Dot product
A = [2, 4, 7]
B = [3, 1, 5]
C = [1, 6, 8]

# Jarak Euclidean antara Vektor A dan Vektor B
i =0
eucli_AB= (sum((A[i] - B[i]) ** 2 for i in range(len(A)))) ** 0.5
print(f"a) Jarak Euclidean : {eucli_AB}")

# Jarak Manhattan antara Vektor A dan Vektor B.
man_AB = sum(abs(A[i] - B[i]) for i in range(len(A)))
print(f"\nb) Jarak Manhattan : {man_AB}")

# Jarak Minkowski antara Vektor A dan Vektor B.
p = 3
minkow_AB =(sum(abs(A[i] - B[i]) ** p for i in range(len(A)))) ** (1/p)
print(f"\nc) Jarak Minkowski (P = 3) : {minkow_AB}")

#  Dot product (produk dot) antara Vektor B dan Vektor C & Vektor A dan Vektor B
dot_AB = sum(A[i] * B[i] for i in range(len(A)))
dot_BC = sum(B[i] * C[i] for i in range(len(B)))

print(f"\nd) Nilai dot vektor A dan B : {dot_AB}, Nilai dot vektor B dan C : {dot_BC}")

a) Jarak Euclidean : 3.7416573867739413

b) Jarak Manhattan : 6

c) Jarak Minkowski (P = 3) : 3.3019272488946263

d) Nilai dot vektor A dan B : 45, Nilai dot vektor B dan C : 49


#4. Bandingkan hasil perhitungan jarak di atas dan berikan analisis tentang perbedaan hasil yang Anda temukan.

    Pada dasarnya setiap rumus sama saja hanya memiliki perbedaan pada nilai p, dimana Euclid memiliki p = 2, dan manhattan p = 1. Sehingga dengan begitu sebenarnya ketiga perhitungan menggunakan rumus minkowski dengan nilai p nya saja yang berbeda. Jarak Euclidean memberikan hasil yang lebih kecil dari jarak Manhattan dan jarak Minkowski, karena jarak Euclidean mengukur kedekatan dua vektor berdasarkan sudut antara kedua vektor tersebut, sedangkan jarak Manhattan dan jarak Minkowski mengukur kedekatan dua vektor berdasarkan jarak yang harus ditempuh untuk bergerak dari satu vektor ke vektor lainnya.
    
#5. Hitunglah dot product (produk dot) antara Vektor B dan Vektor C. Tentukan apakah dot product antara Vektor A dan Vektor B adalah positif, negatif, atau nol. Jelaskan arti dari hasil yang Anda dapatkan.

    Hasil positif pada nilai dot menunjukan bahwa kedua vektor memiliki korelasi yang baik, karena memiliki arah yang semakin berdekatan. Sedangkan negarif memiliki arti berlawanan dan 0 memberikan indikasi tegaklurus pada kedua vektor. Pada dot product vektor A dan B memiliki nilai positif sehingga memiliki kedekatan yang baik.
    
#6. Kesimpulannya, apa yang dapat Anda simpulkan dari perhitungan jarak vektor antara Vektor A, B, dan C? Apa yang dapat Anda simpulkan dari hasil dot product antara Vektor B dan C dalam konteks analisis data?

    Perhitungan jarak euclid, manhattan, dan minkowski mencari jarak antara kedua vektor. Dengan begitu semakin kecil jaraknya semakin similiar komponen dari kedua vektor. Sedangkan dot product dapat memberikan arah serta korelasi dari data berbentuk vektor. Pada perhitungan jarak vektor A, B, C dapat dilihat bahwa data cukup bervariasi. Kemudian pada dot product B dan C memiliki nilai 49, yang berarti kedua vektor memiliki arah yang lumayan sama dan cukup terkorelasi.

In [5]:
# III. Simple Linear Regression and Multiple Linear Regression
X1 = [1500, 2000, 1200, 1800, 1600]
X2 = [3, 4, 2, 3, 2]
Y = [250, 300, 180, 220, 200]

X1_mean = (1500+2000+1200+1800+1600)/5
X2_mean = (3+4+2+3+2)/5
Y_mean = (250+300+180+220+200)/4
print("X1 mean\t: ", X1_mean)
print("X2 mean\t: ", X2_mean)
print("Y mean\t: ", Y_mean)

X1 mean	:  1620.0
X2 mean	:  2.8
Y mean	:  287.5


In [6]:
#Deviation
X1_deviation = [X1[0] - X1_mean, X1[1] - X1_mean, X1[2] - X1_mean, X1[3] - X1_mean, X1[4] - X1_mean]
X2_deviation = [X2[0] - X2_mean, X2[1] - X2_mean, X2[2] - X2_mean, X2[3] - X2_mean, X2[4] - X2_mean]
Y_deviation = [Y[0] - Y_mean, Y[1] - Y_mean, Y[2] - Y_mean, Y[3] - Y_mean, Y[4] - Y_mean]
print("X1 deviation\t: ", X1_deviation)
print("X2 deviation\t: ", X2_deviation)
print("Y deviation\t: ", Y_deviation)

X1_Y_deviation_product = (X1_deviation[0] * Y_deviation[0] + X1_deviation[1] * Y_deviation[1] + X1_deviation[2] * Y_deviation[2]
                        + X1_deviation[3] * Y_deviation[3] + X1_deviation[4] * Y_deviation[4])
X2_Y_deviation_product = (X2_deviation[0] * Y_deviation[0] + X2_deviation[1] * Y_deviation[1] + X2_deviation[2] * Y_deviation[2]
                        + X2_deviation[3] * Y_deviation[3] + X2_deviation[4] * Y_deviation[4])
print("X1Y product deviation\t: ", X1_Y_deviation_product)

# Product Deviation Square
X1_deviation_square_sum = (X1_deviation[0] * X1_deviation[0] + X1_deviation[1] * X1_deviation[1] + X1_deviation[2] * X1_deviation[2]
                            + X1_deviation[3] * X1_deviation[3] + X1_deviation[4] * X1_deviation[4])
Y_deviation_square_sum = (Y_deviation[0] * Y_deviation[0] + Y_deviation[1] * Y_deviation[1] + Y_deviation[2] * Y_deviation[2]
                        + Y_deviation[3] * Y_deviation[3] + Y_deviation[4] * Y_deviation[4])
X2_deviation_square_sum = (X2_deviation[0] * X2_deviation[0] + X2_deviation[1] * X2_deviation[1] + X2_deviation[2] * X2_deviation[2]
                            + X2_deviation[3] * X2_deviation[3] + X2_deviation[4] * X2_deviation[4])
print("X1 Product Deviation Square\t: ", X1_deviation_square_sum)

X1 deviation	:  [-120.0, 380.0, -420.0, 180.0, -20.0]
X2 deviation	:  [0.20000000000000018, 1.2000000000000002, -0.7999999999999998, 0.20000000000000018, -0.7999999999999998]
Y deviation	:  [-37.5, 12.5, -107.5, -67.5, -87.5]
X1Y product deviation	:  44000.0
X1 Product Deviation Square	:  368000.0


In [7]:
# Hitunglah persamaan regresi linear sederhana dengan X1 sebagai prediktor
b1 = (X1_Y_deviation_product / X1_deviation_square_sum)
print("Slope\t: ", b1)

b0 = (Y_mean - (b1 * X1_mean))
print("Intercept\t: ", b0)

#Fungsi Regresi
YX1_Pred = [b0 + b1 * X1[0], b0 + b1 * X1[1], b0 + b1 * X1[2], b0 + b1 * X1[3], b0 + b1 * X1[4]]

Slope	:  0.11956521739130435
Intercept	:  93.80434782608694


In [8]:
#Tentukan koefisien determinasi (R-squared) dari model regresi.
SS_total = Y_deviation_square_sum
SS_error = ((YX1_Pred[0] - Y[0]) ** 2 + (YX1_Pred[1] - Y[1]) ** 2 + (YX1_Pred[2] - Y[2]) ** 2 + (YX1_Pred[3] - Y[3]) ** 2 
+ (YX1_Pred[4] - Y[4]) ** 2)
r_squared = 1 - (SS_error/SS_total)
print("Nilai R Squared : ", r_squared)

Nilai R Squared :  0.20768298308284838


In [9]:
#Jika luas tanah (X1) adalah 1400 meter persegi, berapakah perkiraan harga penjualan rumahnya?
ifX1 = 1400
YifX1_Pred = b0 + b1 * ifX1
print("harga penjualan rumah jika luas tanah (X1) adalah 1400 meter persegi :", YifX1_Pred)

harga penjualan rumah jika luas tanah (X1) adalah 1400 meter persegi : 261.195652173913


In [10]:
#Hitunglah persamaan regresi linear ganda dengan X1 dan X2 sebagai prediktor.
b1 = (X1_Y_deviation_product / X1_deviation_square_sum)
print("Slope X1\t: ", b1)

b2 = (X2_Y_deviation_product / X2_deviation_square_sum)
print("Slope X2\t: ", b2)

b0_X1X2 = (Y_mean - (b1 * X1_mean) - (b2 * X2_mean))
print("Intercept\t: ", b0_X1X2)

#Fungsi Regresi
YX1X2_Pred = [b0_X1X2 + (b1 * X1[0]) + (b2 * X2[0]), b0_X1X2 + b1 * X1[1] + b2 * X2[1], b0_X1X2 + b1 * X1[2] + b2 * X2[2]
            , b0_X1X2 + b1 * X1[3] + b2 * X2[3], b0_X1X2 + b1 * X1[4] + b2 * X2[4]]
print(YX1_Pred)
print(YX1X2_Pred)
print(Y_mean)

Slope X1	:  0.11956521739130435
Slope X2	:  53.571428571428555
Intercept	:  -56.195652173913004
[273.1521739130435, 332.9347826086956, 237.28260869565216, 309.02173913043475, 285.1086956521739]
[283.86645962732916, 397.22049689440996, 194.42546583850933, 319.73602484472053, 242.25155279503107]
287.5


In [11]:
#Tentukan koefisien determinasi (R-squared) dari model regresi.
SS_total2 = Y_deviation_square_sum
SS_error2 = ((YX1X2_Pred[0] - Y[0]) ** 2 + (YX1X2_Pred[1] - Y[1]) ** 2 + (YX1X2_Pred[2] - Y[2]) ** 2 + (YX1X2_Pred[3] - Y[3]) ** 2 
+ (YX1X2_Pred[4] - Y[4]) ** 2)
r_squared2 = 1 - (SS_error2/SS_total2)
print("Nilai R Squared : ", r_squared2)

Nilai R Squared :  0.11021664804691833


In [12]:
#Jika luas tanah (X1) adalah 1400 meter persegi dan jumlah kamar tidur (X2) adalah 2, 
#berapakah  perkiraan harga penjualan rumahnya?

ifX1 = 1400
ifX2 = 2
YifX1X2_Pred = b0_X1X2 + (b1 * ifX1) + (b2 * ifX2)
print("harga penjualan rumah jika luas tanah (X1) adalah 1400 meter persegi\n, dan jumlah kamar tidur (X2) adalah 2 :", YifX1X2_Pred)

harga penjualan rumah jika luas tanah (X1) adalah 1400 meter persegi
, dan jumlah kamar tidur (X2) adalah 2 : 218.3385093167702


#7. Bandingkan hasil dari regresi linear sederhana (Bagian A) dan regresi linear ganda (Bagian B). Apa kesimpulan yang dapat Anda ambil tentang kualitas model dan faktor-faktor yang memengaruhi harga penjualan rumah?

Pada regresi linear sederhana, nilai r squarednya adalah : 0.20768298308284838. Dengan begitu dapat dilihat bahwa variabel X1 masih cenderung kurang baik dalam menjelaskan variabel Y terkait hubungan antara kedua variabel, keragaman variabel Y, dan tingkat kesesuaian data dengan model regresi.

Pada regresi linear sederhana, nilai r squarednya adalah : 0.11021664804691833. Dengan begitu dapat dilihat bahwa variabel X1 dan X2 sangat kurang baik dalam menjelaskan variabel Y terkait hubungan antara kedua variabel, keragaman variabel Y, dan tingkat kesesuaian data dengan model regresi.

In [13]:
# IV. Decision Tree
X1 = [25, 30, 35, 40, 45]
X2 = [4, 5, 6, 7, 8]
Y = [0, 0, 0, 1, 1]

# Hitung Probabilitas Y
p0 = 3 / 5
p1 = 2 / 5

# Hitung entropi H(Y) 
log2_p0 = -0.7369
log2_p1 = -1.3219
HY = -(p0 * log2_p0 + p1 * log2_p1)

print("entropi (H(Y)) : ", HY)

entropi (H(Y)) :  0.9709


In [14]:
#Hitunglah entropi H(Y|X1)
X1_25_0 = 1
X1_25_1 = 0
X1_30_0 = 1
X1_30_1 = 0
X1_35_0 = 1
X1_35_1 = 0
X1_40_0 = 0
X1_40_1 = 1
X1_45_0 = 0
X1_45_1 = 1

HY_X1_25 = 0
HY_X1_30 = 0
HY_X1_35 = 0
HY_X1_40 = 0
HY_X1_45 = 0

# Hitung entropi target Y berdasarkan fitur X1
HY_X1 = (1/5 * HY_X1_25 + 1/5 * HY_X1_30 + 1/5 * HY_X1_35 + 1/5 * HY_X1_40 + 
        1/5 * HY_X1_45)
print("entropi (H(Y|X1)) : ", HY_X1)

entropi (H(Y|X1)) :  0.0


In [15]:
# Hitunglah entropi (H(Y|X2))
X2_4_0 = 1
X2_4_1 = 0
X2_5_0 = 1
X2_5_1 = 0
X2_6_0 = 1
X2_6_1 = 0
X2_7_0 = 0
X2_7_1 = 1
X2_8_0 = 0
X2_8_1 = 1

HY_X2_4 = 0
HY_X2_5 = 0
HY_X2_6 = 0
HY_X2_7 = 0
HY_X2_8 = 0

# Hitung entropi target Y berdasarkan fitur X2
HY_X2 = (1/5 * HY_X2_4  + 1/5 * HY_X2_5 + 1/5 * HY_X2_6 + 1/5 * HY_X2_7 + 
        1/5 * HY_X2_8 )
print("entropi (H(Y|X2)) : ", HY_X2)

entropi (H(Y|X2)) :  0.0


#4.Apa makna hasil entropi yang Anda hitung dalam konteks pengambilan keputusan?

Entropi memiliki arti ketidak pastian, semakin tinggi nilainya semakin tinggi pula ketidak pastiannya. Namun pada decision tree, entropi dapat mengindikasikan keberagaman probabilitas dan tingkat homogen dari dataset. Dalam decision tree, entropi juga dapat digunakan untuk memilih fitur terbaik. Karena entropi target Y berdasarkan fitur X1 dan X2 : 0.0 menunjukan model decision tree akan dapat memprediksi nilai target Y dengan sangat baik.

In [16]:
# Hitunglah Gini impurity (Gini(Y)) dari target Y pada dataset ini.
Gini_Y = 1 - (p0 ** 2 + p1 ** 2)
print(Gini_Y)

0.48


In [17]:
# Hitunglah Gini impurity (Gini(Y|X1)) dari target Y berdasarkan fitur X1.
print("Gini(Y|X1 = 25) = 1 - (1^2 + 0^2) = 1 - 1 = 0")
print("Gini(Y|X1 = 30) = 1 - (1^2 + 0^2) = 1 - 1 = 0")
print("Gini(Y|X1 = 35) = 1 - (1^2 + 0^2) = 1 - 1 = 0")
print("Gini(Y|X1 = 40) = 1 - (0^2 + 1^2) = 1 - 1 = 0")
print("Gini(Y|X1 = 45) = 1 - (0^2 + 1^2) = 1 - 1 = 0")

print("\nGini(Y|X1) = 0")

Gini(Y|X1 = 25) = 1 - (1^2 + 0^2) = 1 - 1 = 0
Gini(Y|X1 = 30) = 1 - (1^2 + 0^2) = 1 - 1 = 0
Gini(Y|X1 = 35) = 1 - (1^2 + 0^2) = 1 - 1 = 0
Gini(Y|X1 = 40) = 1 - (0^2 + 1^2) = 1 - 1 = 0
Gini(Y|X1 = 45) = 1 - (0^2 + 1^2) = 1 - 1 = 0

Gini(Y|X1) = 0


In [18]:
# Hitunglah Gini impurity (Gini(Y|X1)) dari target Y berdasarkan fitur X1.
print("Gini(Y|X2 = 4) = 1 - (1^2 + 0^2) = 1 - 1 = 0")
print("Gini(Y|X2 = 5) = 1 - (1^2 + 0^2) = 1 - 1 = 0")
print("Gini(Y|X2 = 6) = 1 - (1^2 + 0^2) = 1 - 1 = 0")
print("Gini(Y|X2 = 7) = 1 - (0^2 + 1^2) = 1 - 1 = 0")
print("Gini(Y|X2 = 8) = 1 - (0^2 + 1^2) = 1 - 1 = 0")

print("\nGini(Y|X2) = 0")

Gini(Y|X2 = 4) = 1 - (1^2 + 0^2) = 1 - 1 = 0
Gini(Y|X2 = 5) = 1 - (1^2 + 0^2) = 1 - 1 = 0
Gini(Y|X2 = 6) = 1 - (1^2 + 0^2) = 1 - 1 = 0
Gini(Y|X2 = 7) = 1 - (0^2 + 1^2) = 1 - 1 = 0
Gini(Y|X2 = 8) = 1 - (0^2 + 1^2) = 1 - 1 = 0

Gini(Y|X2) = 0


#8. Apa makna hasil Gini impurity yang Anda hitung dalam konteks pengambilan keputusan?

Semakin rendah gini semakin homogen data tersebut. Nilai gini yang didapat berupa Gini_Y : 0.48, Gini(Y|X1) = 0, dan Gini(Y|X2) = 0. Karena gini digunakan untuk mengukur sejauh mana sebuah atribut efektif dalam proses klasifikasi data ke dalam kelas yang berbeda, maka nilai 0 menunjukan  keadaan "murni", yaitu semua elemen dalam set tersebut termasuk dalam satu kelas tertentu.

#9. Bandingkan hasil entropi dan Gini impurity untuk fitur X1 dan X2. Apa yang dapat Anda simpulkan tentang impurity fitur-fitur ini dalam klasifikasi pelanggan membeli atau tidak?

Dua-duanya bernilai nol, karena tidak ada kelas pada X1 dna X2. Dengan begitu nilai entropi dan gini menjadi kecil dan menunjukan efiktifitas yang tinggi dalam klasifikasi karena setiap nilai diklasifikasi sebagai 1 kelas.

In [19]:
# Hitung Jarak Euclidean
# X1 = 5 X2 = 4
# print("(5-3)^2 + (4-5)^2 = ", (5-3)**2 + (4-5)**2)
# print("(5-4)^2 + (4-6)^2 = ", (5-4)**2 + (4-6)**2)
# print("(5-6)^2 + (4-3)^2 = ", (5-6)**2 + (4-3)**2)
# print("(5-7)^2 + (4-5)^2 = ", (5-7)**2 + (4-5)**2)
# print("(5-9)^2 + (4-2)^2 = ", (5-9)**2 + (4-2)**2)
# print("(5-8)^2 + (4-3)^2 = ", (5-8)**2 + (4-3)**2)
# print("(5-2)^2 + (4-9)^2 = ", (5-2)**2 + (4-9)**2)

# # Jumlahkan
# print("\nTotal = ", 5+5+2+5+20+10+45)

print("((5-3)^2 + (4-5)^2)^0.5 = ", ((5-3)**2 + (4-5)**2)**0.5)
print("((5-4)^2 + (4-6)^2)^0.5 = ", ((5-4)**2 + (4-6)**2)**0.5)
print("((5-6)^2 + (4-3)^2)^0.5 = ", ((5-6)**2 + (4-3)**2)**0.5)
print("((5-7)^2 + (4-5)^2)^0.5 = ", ((5-7)**2 + (4-5)**2)**0.5)
print("((5-9)^2 + (4-2)^2)^0.5 = ", ((5-9)**2 + (4-2)**2)**0.5)
print("((5-8)^2 + (4-3)^2)^0.5 = ", ((5-8)**2 + (4-3)**2)**0.5)
print("((5-2)^2 + (4-9)^2)^0.5 = ", ((5-2)**2 + (4-9)**2)**0.5)

((5-3)^2 + (4-5)^2)^0.5 =  2.23606797749979
((5-4)^2 + (4-6)^2)^0.5 =  2.23606797749979
((5-6)^2 + (4-3)^2)^0.5 =  1.4142135623730951
((5-7)^2 + (4-5)^2)^0.5 =  2.23606797749979
((5-9)^2 + (4-2)^2)^0.5 =  4.47213595499958
((5-8)^2 + (4-3)^2)^0.5 =  3.1622776601683795
((5-2)^2 + (4-9)^2)^0.5 =  5.830951894845301


In [20]:
#k-Nearest Neighbors (kNN) dan Support Vector Machine (SVM)
# Hitung Jarak Euclidean
# X1 = 5 X2 = 4
# print("(5-3)^2 + (4-5)^2 = ", (5-3)**2 + (4-5)**2)
# print("(5-4)^2 + (4-6)^2 = ", (5-4)**2 + (4-6)**2)
# print("(5-6)^2 + (4-3)^2 = ", (5-6)**2 + (4-3)**2)
# print("(5-7)^2 + (4-5)^2 = ", (5-7)**2 + (4-5)**2)
# print("(5-9)^2 + (4-2)^2 = ", (5-9)**2 + (4-2)**2)
# print("(5-8)^2 + (4-3)^2 = ", (5-8)**2 + (4-3)**2)
# print("(5-2)^2 + (4-9)^2 = ", (5-2)**2 + (4-9)**2)

# # Jumlahkan
# print("\nTotal = ", 5+5+2+5+20+10+45)

print("((5-3)^2 + (4-5)^2)^0.5 = ", ((5-3)**2 + (4-5)**2)**0.5)
print("((5-4)^2 + (4-6)^2)^0.5 = ", ((5-4)**2 + (4-6)**2)**0.5)
print("((5-6)^2 + (4-3)^2)^0.5 = ", ((5-6)**2 + (4-3)**2)**0.5)
print("((5-7)^2 + (4-5)^2)^0.5 = ", ((5-7)**2 + (4-5)**2)**0.5)
print("((5-9)^2 + (4-2)^2)^0.5 = ", ((5-9)**2 + (4-2)**2)**0.5)
print("((5-8)^2 + (4-3)^2)^0.5 = ", ((5-8)**2 + (4-3)**2)**0.5)
print("((5-2)^2 + (4-9)^2)^0.5 = ", ((5-2)**2 + (4-9)**2)**0.5)

((5-3)^2 + (4-5)^2)^0.5 =  2.23606797749979
((5-4)^2 + (4-6)^2)^0.5 =  2.23606797749979
((5-6)^2 + (4-3)^2)^0.5 =  1.4142135623730951
((5-7)^2 + (4-5)^2)^0.5 =  2.23606797749979
((5-9)^2 + (4-2)^2)^0.5 =  4.47213595499958
((5-8)^2 + (4-3)^2)^0.5 =  3.1622776601683795
((5-2)^2 + (4-9)^2)^0.5 =  5.830951894845301


Pilih 3 Titik  :
Tetangga 1: (6, 3) dengan jarak ≈ 1.414  
Tetangga 2: (3, 5) dengan jarak ≈ 2.236  
Tetangga 3: (4, 6) dengan jarak ≈ 2.236

Menentukan Titik :  
(6, 3) = B  
(3, 5) = A  
(4, 6) = A  

Jadi, titik (5, 4) diklasifikasikan sebagai A.

In [21]:
#Naive Bayes Classifier
email_text = ['Dapatkan produk super diskon!', 'Peningkatan performa di tempat kerja anda!', 
              'Promo spesial hanya hari ini!', 'Rapat penting di ruang konferensi.', 'Anda memenangkan hadiah besar!']
Label = [1, 0, 1, 0, 1]

In [57]:
#Hitunglah probabilitas prior (P(Spam) dan P(Bukan Spam)) berdasarkan jumlah email spam dan bukan spam dalam dataset.
p_spam = 3/5
p_nospam = 2/5

print("P(spam) : ", p_spam)
print("P(nospam) : ", p_nospam)

#Hitunglah probabilitas likelihood P(Kata | Spam) dan P(Kata | Bukan Spam) untuk kata "produk" dan "hadiah."
p_produk_spam = 1/3
p_hadiah_spam = 1/3
p_produk_nospam = 0
p_hadiah_nospam = 0

print("P(produk|spam) : ", p_produk_spam)
print("P(produk|nospam) : ", p_produk_nospam)
print("P(hadiah|spam) : ", p_hadiah_spam)
print("P(hadiah|nospam) : ", p_hadiah_nospam)

#Gunakan metode Naive Bayes untuk mengklasifikasikan email "Promo spesial hanya hari ini!" apakah spam atau bukan spam.
email = "Promo spesial hanya hari ini!"
p_promo_spam = 1/3
p_spesial_spam = 1/3
p_hanya_spam = 1/3
p_hari_spam = 1/3
p_ini_spam = 1/3

p_promo_nospam = 0
p_spesial_nospam = 0
p_hanya_nospam = 0
p_hari_nospam = 0
p_ini_nospam = 0

p_email_spam = p_spam * p_promo_spam * p_spesial_spam * p_hanya_spam * p_hari_spam * p_ini_spam
p_email_nospam = p_nospam * p_promo_nospam * p_spesial_nospam * p_hanya_nospam * p_hari_nospam * p_ini_nospam

print("P(Promo spesial hanya hari ini!|spam) : ", p_email_spam)
print("P(Promo spesial hanya hari ini!|nospam) : ", p_email_nospam)

#Karena P(Promo spesial hanya hari ini!|spam) > P(Promo spesial hanya hari ini!|nospam) 
#Maka email tersebut dianggap sebagai spam

P(spam) :  0.6
P(nospam) :  0.4
P(produk|spam) :  0.3333333333333333
P(produk|nospam) :  0
P(hadiah|spam) :  0.3333333333333333
P(hadiah|nospam) :  0
P(Promo spesial hanya hari ini!|spam) :  0.002469135802469135
P(Promo spesial hanya hari ini!|nospam) :  0.0


#Jelaskan apa yang dimaksud dengan "Naive" dalam Naive Bayes Classifier dan mengapa metode ini dapat diklasifikasikan sebagai metode probabilistik.

Karena untuk pengambilan keputusan ditentukan berdasarkan peluang. Sehingga probabilistik menjadi nilai utama dalam metode ini.

In [1]:
!jupyter nbconvert --to html "./IF540_E_Laporan_01_58511_RagilMuhamadRamadhan.ipynb" --output-dir="./"

[NbConvertApp] Converting notebook ./IF540_E_Laporan_01_58511_RagilMuhamadRamadhan.ipynb to html
[NbConvertApp] Writing 682110 bytes to IF540_E_Laporan_01_58511_RagilMuhamadRamadhan.html
